### PR TITLE
Serializing an empty array should maintain the root key

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Relates to: #1153

### CODE REVIEW

File: lib/active_model/serializer/array_serializer.rb [hunk 1]
```diff
@@ -27,9 +28,18 @@ def initialize(resources, options = {})
...
       def json_key
         key = root || @serializers.first.try(:json_key) || object.try(:name).try(:underscore)
+        key ||= inferred_json_key
         key.try(:pluralize)
       end
 
+      def inferred_json_key
+        serializer = options[:serializer]
+        return unless serializer
+        potential_model_name = serializer.name.sub('Serializer', '')
```
Comment (bf4) [inline lib/active_model/serializer/array_serializer.rb]: `serializer.name.remove('Serializer')` :shrug:

**Comment by bf4** [comment]
Let's hold off on merging this till we clarify which arguments, `serializer` vs. `each_serializer` we should be passing to manually specify the serializer for a non-compliant resource (i.e. would fail the Lint).

See my comment in the OP's issue: https://github.com/rails-api/active_model_serializers/issues/1153#issuecomment-140624893

**Comment by NullVoxPopuli** [comment]
@bf4  for an empty object (Array, or Hash) - would it matter which serializer vs each_serializer we're using? there isn't really anything to serialize aside from the root key.

it would either be

```
{
  root: {}
}
```
or
```
{
  roots: []
}
```

**Comment by bf4** [comment]
So, AMS doesn't handle primitives, and when they find their way it, they are passed through as is. ( Specifically, when you're not rendering an object with a defined serializer.) 

How options are parsed

1. `serializable_resource = ActiveModel::SerializableResource.new(resource, options)`
1. `if serializable_resource.serializer?`
  - Where `serializer?` is `use_adapter? && !!(serializer)`
    - Where `use_adapter?`: 'True when no explicit adapter given, or explicit appear is truthy (non-nil); False when explicit adapter is falsy (nil or false)'
    - Where `serializer`:
      1. from explicit `:serializer` option, else
      2. implicitly from resource `ActiveModel::Serializer.serializer_for(resource)`
   - The `:serializer` option is removed from the serializer_opts hash
   - If the `:each_serializer` option is present, it is removed from the serializer_opts hash and set as the `:serializer` option
1. The serializer and adapter are created as
  1. `serializer_instance = serializer.new(resource, serializer_opts)`
  2. `adapter_instance = ActiveModel::Serializer::Adapter.create(serializer_instance, adapter_opts)`
1. If the `serializer_instance` was a `ArraySerializer` and the `:serializer` serializer_opts is present, then that serializer is passed into each resource: https://github.com/rails-api/active_model_serializers/blob/a54d237e2828fe6bab1ea5dfe6360d4ecc8214cd/lib/active_model/serializer/array_serializer.rb#L14-L16

The upshot of which is that
- for a collection, the collection serializer is the `:serializer` option and `:each_serializer` is used as the serializer for each resource in the collection
- for a single resource, the `:serializer` option is the resource serializer

**Comment by NullVoxPopuli** [comment]
that is a fantastic explanation!

thanks -- I'll close this PR.

**Pull request closed without merging**